src/time-to-interactive/TTIBoomerang.ts

```diff
@@ -0,0 +1,331 @@
+                ) {
+                    // If timed out, check if any visually ready signals are present, else can't find tti
+                    if (
+                        timeIntervals * this.CHECK_PERIOD >
+                        this.VISUALLY_READY_RESOLVE_TIMEOUT
+                    ) {
+                        if (
+                            !this.lcpTime &&
+                            !this.domContentLoadedEventEnd &&
+                            !this.fcpTime
+                        ) {
+                            // No visually ready timestamps so can't compute tti
+                            this.ttiResolved = true;
+                            this.ttiTracker = {};
+                            reject(
+                                'Insufficient visually ready timestamps to compute TTI'
+                            );
+                        }
+                    }
+                    // Visually ready so start checking for TTI
+                    clearInterval(visuallyReadyInterval);
+
+                    this.visuallyReadyTimestamp = Math.max(
+                        this.fcpTime ? this.fcpTime : 0,
+                        this.lcpTime ? this.lcpTime : 0,
+                        this.domContentLoadedEventEnd
+                            ? this.domContentLoadedEventEnd
+                            : 0
+                    );
+                    resolve(this.visuallyReadyTimestamp);
+                }
+                timeIntervals += 1;
+            }, this.CHECK_PERIOD);
+        });
+    }
+
+    private handleWebVitals(metric: Metric) {
+        if (metric.name === FCP) {
+            metric.entries.forEach(
+                (entry) => (this.fcpTime = entry.startTime + entry.duration)
+            );
+        }
+        if (metric.name === LCP) {
+            metric.entries.forEach(
+                (entry) => (this.lcpTime = entry.startTime + entry.duration)
+            );
+        }
+    }
+
+    computeTTI(visuallyReadyTimestamp: number): Promise<number> {
+        return new Promise<number>((resolve, reject) => {
+            const startBucket = Math.max(
+                this.computeTimeWindow(visuallyReadyTimestamp),
+                0
+            );
+            let currBucket = startBucket;
+            let okIntervals = 0;
+            let totalIntervals = 0;
+
+            const tti: any = setInterval(() => {
+                if (
+                    totalIntervals * this.CHECK_PERIOD >
+                    this.TTI_RESOLVE_TIMEOUT
+                ) {
+                    // TTI did not resolve in timeout period, so TTI can't be computed
+                    this.ttiResolved = true;
+                    this.ttiTracker = {};
+                    reject('TTI computation timed out');
+                }
+                const endBucket = this.computeTimeWindow();
+                for (let bucket = currBucket; bucket <= endBucket; bucket++) {
```

Comment:
suggestion: This has a lot of nesting, and cyclomatic complexity. Could you clean it up? Helper functions would make this easier to read.

Reply:
cleaned up in refactor